ksqldb-parser/src/test/java/io/confluent/ksql/parser/VariableSubstitutorTest.java

```diff
@@ -0,0 +1,221 @@
+        Pair.of("DESCRIBE ${quotedIdentifier2};", "Got: '\"EXTENDED _id_\"'"),
+        Pair.of("DESCRIBE ${backQuotedIdentifier2};", "Got: '`EXTENDED _id_`'"),
+        Pair.of("DESCRIBE ${singleQuoteIdentifier2};", "Got: ''EXTENDED _id_''")
+    );
+
+    assertThrowOnInvalidVariables(statements, variablesMap);
+  }
+
+  @Test
+  public void shouldSubstituteVariableOnInsert() {
+    // Given
+    final Map<String, String> variablesMap = new ImmutableMap.Builder<String, String>() {{
+      put("identifier", "_id_");
+      put("num", "1");
+      put("dec", "0.32");
+      put("bool", "false");
+      put("str", "'john'");
+    }}.build();
+
+    final List<Pair<String, String>> statements = Arrays.asList(
+        // INSERT VALUES
+        Pair.of("INSERT INTO ${identifier} VALUES (${num}, ${bool}, ${dec}, ${str});",
+            "INSERT INTO _id_ VALUES (1, false, 0.32, 'john');")
+    );
+
+    // When/Then
+    assertReplacedStatements(statements, variablesMap);
+  }
+
+  @Test
+  public void shouldSubstituteVariableOnCreate() {
+    // Given
+    final Map<String, String> variablesMap = new ImmutableMap.Builder<String, String>() {{
+      put("identifier", "_id_");
+      put("quotedIdentifier", "\"_id_\"");
+      put("backQuotedIdentifier", "`_id_`");
+      put("topicName", "'name1'");
+      put("replicas", "3");
+    }}.build();
+
+    final List<Pair<String, String>> statements = Arrays.asList(
+        // CREATE STREAM
+        Pair.of("CREATE STREAM ${identifier} WITH (kafka_topic=${topicName}, replicas=${replicas});",
+            "CREATE STREAM _id_ WITH (kafka_topic='name1', replicas=3);"),
+
+        Pair.of("CREATE STREAM ${quotedIdentifier} WITH (kafka_topic=${topicName}, replicas=${replicas});",
+            "CREATE STREAM \"_id_\" WITH (kafka_topic='name1', replicas=3);"),
+
+        Pair.of("CREATE STREAM ${backQuotedIdentifier} WITH (kafka_topic=${topicName}, replicas=${replicas});",
+            "CREATE STREAM `_id_` WITH (kafka_topic='name1', replicas=3);"),
+
+        Pair.of("CREATE STREAM ${backQuotedIdentifier} WITH (kafka_topic=${topicName}, replicas=${replicas});",
+            "CREATE STREAM `_id_` WITH (kafka_topic='name1', replicas=3);")
+    );
+
+    // When/Then
+    assertReplacedStatements(statements, variablesMap);
+  }
+
+  @Test
+  public void shouldSanitizeStatements() {
+    // Given
+    final Map<String, String> variablesMap = new ImmutableMap.Builder<String, String>() {{
+      put("escapeQuote", "'t1', value_format='AVRO'");
+    }}.build();
+
+    final List<Pair<String, String>> statements = Arrays.asList(
+        // CREATE
+        Pair.of("CREATE STREAM s1 WITH (kafka_topic=${escapeQuote});",
+            "CREATE STREAM s1 WITH (kafka_topic='t1'', value_format=''AVRO');")
+    );
+
+    // When/Then
+    assertReplacedStatements(statements, variablesMap);
+  }
+
+  @Test
+  public void shouldThrowOnSQLInjection() {
```

Comment:
This works (identifying sql injection) because variable names must not contain spaces?

Reply:
Variable names do not contain spaces, but values can. However, it all depends on how to reference the values in the statement. All values must be enclosed into single-quotes, but single-quotes are not part of the value in the statement.

For instance:
```
// This should throw an error because ${id} has spaces.
ksql> define id = '5 and id != 5';
ksql> SELECT * FROM s1 WHERE id = ${id};

// This is a valid statement. The query will print | 5 and id != 5 | in column 1 because it is referenced inside the single-quotes, which defines the column as String.
ksql> define id = '5 and id != 5';
ksql> SELECT '${id}' FROM ...
```